File: ch_gpio/adc.c
#include <stdio.h>
#include <wiringPi.h>
#include <wiringPiSPI.h>
#include <unistd.h>

int readadc(int ch)
{
    int errno;
    unsigned char data[3];
    unsigned int adcout;
    wiringPiSPISetup(0, 500000);

    data[0] = 1; 
    data[1] = (8 + ch) << 4; 
    data[2] = 0; 

    errno = wiringPiSPIDataRW(0, data, 3);

    adcout = ((data[1] & 0x03) << 8) + data[2];

    return adcout;
}

int main(int argc, char *argv[])
{
    int ch;
    wiringPiSetupPhys();
    while(1){
        for(ch = 0; ch <= 1; ch++){
            int rst = readadc(ch);
            printf("ch %d is %d\n", ch, rst);
        }
        sleep(1);
    }

    return 0;
}

File: ch_gpio/adc.py
#!/usr/bin/env python

import spidev
import time

spi = spidev.SpiDev()
spi.open(0, 0)

def readadc(ch):
    if ch > 7 or ch < 0:
        return -1
    rst = spi.xfer2([1, (8 + ch) << 4, 0])
    adcout = ((rst[1] & 0x03) << 8) + rst[2]
    return adcout

while True:
    for i in range(0, 2):
        print("Ch " + str(i) + " = " + str(readadc(i)))
    time.sleep(1)


